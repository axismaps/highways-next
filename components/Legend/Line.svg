<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 17.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 width="60px" height="20px" viewBox="0 20 80 40" enable-background="new 0 20 80 40" xml:space="preserve">
<path fill="currentColor" stroke="#999" stroke-miterlimit="10" d="M6.4,54.7c-1.4,0-2.7-0.5-3.6-1.4c-1-0.9-1.5-2.1-1.5-3.4
	c0-1.3,0.6-2.5,1.6-3.3l23.5-20c0.9-0.8,2.2-1.2,3.4-1.2c1.3,0,2.6,0.5,3.6,1.4l18.3,16.7L70,26.7c1-0.9,2.2-1.4,3.6-1.4
	c1.3,0,2.5,0.5,3.5,1.3c1,0.9,1.6,2.1,1.6,3.3c0,1.3-0.5,2.5-1.5,3.4l-21.8,20c-0.9,0.8-2.2,1.3-3.6,1.3c-1.4,0-2.7-0.5-3.6-1.3
	L29.8,36.5l-19.9,17C8.9,54.2,7.7,54.7,6.4,54.7z"/>
</svg>
