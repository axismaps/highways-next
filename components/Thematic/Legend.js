import React from 'react';
import PropTypes from 'prop-types';
import { Box, Grid, Heading } from '@chakra-ui/react';

const Legend = ({ title, scale, colors }) => (
  <Box>
    <Heading as="h3" size="sm" mt={6} mb={2}>
      {title}
    </Heading>
    <Grid templateColumns={`repeat(${colors.length}, 1fr)`}>
      {colors.map((color, i) => (
        <Box key={color} bg={color} h="1rem">
          <Box
            fontSize={12}
            float="right"
            pos="relative"
            top="15px"
            left="15px"
            w="30px"
            textAlign="center"
          >
            {`${scale[i] || ''}`}
          </Box>
        </Box>
      ))}
    </Grid>
  </Box>
);

Legend.propTypes = {
  title: PropTypes.string.isRequired,
  scale: PropTypes.arrayOf(PropTypes.number).isRequired,
  colors: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Legend;
