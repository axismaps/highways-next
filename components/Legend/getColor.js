import { last, findLast } from 'lodash';
import { rgb } from 'd3';
import Line from './Line.svg';
import style from '../Atlas/style.json';

const getLayerStyle = (layer, type) => {
  let layerStyle = style.layers.filter(
    l =>
      l.filter &&
      l['source-layer'] === layer.name &&
      l.filter.find(f => Array.isArray(f) && f[2] && f[2][0] === type)
  );
  if (!layerStyle.length) layerStyle = style.layers.filter(l => l['source-layer'] === layer.name);
  return layerStyle;
};

const getColor = (layer, type) => {
  const layerStyle = getLayerStyle(layer, type);
  if (!layerStyle.length) return { backgroundColor: 'white' };
  if (layerStyle[0].paint['fill-color']) {
    let backgroundColor = layerStyle[0].paint['fill-color'];
    let borderColor = layerStyle[0].paint['fill-outline-color'];
    if (!borderColor)
      borderColor = layerStyle[1]
        ? layerStyle[1].paint['fill-color']
        : layerStyle[0].paint['fill-color'];

    [backgroundColor, borderColor] = [backgroundColor, borderColor].map(bColor => {
      if (!bColor) return null;
      let color = bColor;
      if (Array.isArray(color)) color = last(color);
      if (color.match(/^hsl/gi)) color = rgb(color).formatHex();
      return color;
    });
    return {
      borderColor,
      backgroundColor,
      borderWidth: 2,
      m: 1,
    };
  }
  if (layerStyle[0].paint['line-color']) {
    return {
      as: Line,
      color: findLast(layerStyle, lStyle => lStyle.paint['line-color']).paint['line-color'],
      w: '35px',
      ml: '5px',
    };
  }
  return { backgroundColor: 'white' };
};

export default getColor;
