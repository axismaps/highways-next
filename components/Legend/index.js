import React from 'react';
import PropTypes from 'prop-types';
import useSWR from 'swr';
import axios from 'axios';
import { Box, Grid, Heading, Text } from '@chakra-ui/react';

import getColor from './getColor';
import Loading from '../Loading';
import useDebounce from '../../utils/useDebounce';

const fetcher = url => axios.get(url).then(({ data }) => data);

const Legend = ({ year }) => {
  const debouncedYear = useDebounce(year, 500);
  const { data: layers, error } = useSWR(
    debouncedYear ? `${process.env.NEXT_PUBLIC_SEARCH_API}/layers?year=${debouncedYear}` : null,
    fetcher
  );

  if (!layers || error) return <Loading />;

  return (
    <Box>
      <Heading size="md" mb={2}>
        Legend
      </Heading>
      <Box px="15px" pb="10px" pt="1px" backgroundColor="white" borderRadius="5px" boxShadow="md">
        {layers.map(layer => (
          <Box key={layer.id}>
            <Heading size="sm" mt={4} mb={2}>
              {layer.title}
            </Heading>
            <Grid templateColumns="45px 1fr" columnGap="10px" rowGap="10px">
              {layer.types.map(type => (
                <React.Fragment key={type}>
                  <Box {...getColor(layer, type)} />
                  <Text>{type}</Text>
                </React.Fragment>
              ))}
            </Grid>
          </Box>
        ))}
      </Box>
    </Box>
  );
};

Legend.propTypes = {
  year: PropTypes.number.isRequired,
};

export default Legend;
