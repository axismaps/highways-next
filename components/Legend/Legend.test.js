import React from 'react';
import axios from 'axios';
import renderer from 'react-test-renderer';
import MockAdapter from 'axios-mock-adapter';
import { render, screen } from '../../test/test-utils';

import Legend from './index';
import layers from '../../__mocks__/layers.data';

describe('Legend', () => {
  const mock = new MockAdapter(axios);
  mock.onGet(`${process.env.NEXT_PUBLIC_SEARCH_API}/layers?year=1950`).reply(200, layers);

  const legendProps = {
    year: 1950,
    layerHandler: () => {},
  };

  it('renders for 1950 without crashing', async () => {
    render(<Legend {...legendProps} />);
    await screen.findByText('Legend');
    expect(screen.getByRole('heading', { name: 'Legend' })).toBeInTheDocument();
    expect(screen.getByText('Roads')).toBeInTheDocument();
  });

  it('matches snapshot', async () => {
    render(<Legend {...legendProps} />);
    await screen.findByText('Legend');

    const tree = renderer.create(<Legend {...legendProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
