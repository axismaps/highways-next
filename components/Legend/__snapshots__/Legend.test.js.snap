// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend matches snapshot 1`] = `
<div
  className="css-0"
>
  <h2
    className="chakra-heading css-wlea3r"
  >
    Legend
  </h2>
  <div
    className="css-14z9jvq"
  >
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Buildings
      </h2>
      <div
        className="css-3isub8"
      >
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Commercial
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Corporate Office
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Health
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Industrial
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Park Facility
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Public and Institutional
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Religious
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Residential
        </p>
        <div
          className="css-1aabom1"
        />
        <p
          className="chakra-text css-0"
        >
          Transportantion and Utility
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Roads
      </h2>
      <div
        className="css-3isub8"
      >
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Highway
        </p>
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Other
        </p>
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Primary
        </p>
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Residential
        </p>
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Secondary
        </p>
        <svg-mock
          className="css-763syg"
        />
        <p
          className="chakra-text css-0"
        >
          Tertiary
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Utilities
      </h2>
      <div
        className="css-3isub8"
      >
        <svg-mock
          className="css-e9nj70"
        />
        <p
          className="chakra-text css-0"
        >
          Canal
        </p>
        <svg-mock
          className="css-e9nj70"
        />
        <p
          className="chakra-text css-0"
        >
          Drain
        </p>
        <svg-mock
          className="css-8hv0vx"
        />
        <p
          className="chakra-text css-0"
        >
          Railroad
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Hidrography
      </h2>
      <div
        className="css-3isub8"
      >
        <svg-mock
          className="css-e9nj70"
        />
        <p
          className="chakra-text css-0"
        >
          Bayou
        </p>
        <svg-mock
          className="css-e9nj70"
        />
        <p
          className="chakra-text css-0"
        >
          River
        </p>
        <svg-mock
          className="css-e9nj70"
        />
        <p
          className="chakra-text css-0"
        >
          Stream
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Open Spaces
      </h2>
      <div
        className="css-3isub8"
      >
        <div
          className="css-bfh33g"
        />
        <p
          className="chakra-text css-0"
        >
          Field
        </p>
        <div
          className="css-bfh33g"
        />
        <p
          className="chakra-text css-0"
        >
          Golf Course
        </p>
        <div
          className="css-gihtsx"
        />
        <p
          className="chakra-text css-0"
        >
          Graveyard
        </p>
        <div
          className="css-gihtsx"
        />
        <p
          className="chakra-text css-0"
        >
          Park
        </p>
        <div
          className="css-1ep5i2v"
        />
        <p
          className="chakra-text css-0"
        >
          Square
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Hidrography
      </h2>
      <div
        className="css-3isub8"
      >
        <div
          className="css-p8iwet"
        />
        <p
          className="chakra-text css-0"
        >
          Reservoir
        </p>
        <div
          className="css-1ylr6zp"
        />
        <p
          className="chakra-text css-0"
        >
          Wetland
        </p>
      </div>
    </div>
    <div
      className="css-0"
    >
      <h2
        className="chakra-heading css-1oiljpd"
      >
        Boundaries
      </h2>
      <div
        className="css-3isub8"
      >
        <div
          className="css-7n8q7e"
        />
        <p
          className="chakra-text css-0"
        >
          City Limit
        </p>
        <div
          className="css-12toqkv"
        />
        <p
          className="chakra-text css-0"
        >
          County Limit
        </p>
        <div
          className="css-12toqkv"
        />
        <p
          className="chakra-text css-0"
        >
          County Limits
        </p>
        <div
          className="css-12toqkv"
        />
        <p
          className="chakra-text css-0"
        >
          Subwatershed
        </p>
        <div
          className="css-12toqkv"
        />
        <p
          className="chakra-text css-0"
        >
          Watershed
        </p>
      </div>
    </div>
  </div>
</div>
`;
